### ISSUE TITLE
Explore simplifying dev container definition with a concept of 'container features'

### ISSUE THREAD

**Issue by chrmarti** [microsoft/vscode-remote-release#5442]
The goal of this feature is to address the complexity of adding simple tools to an existing dev container configuration. Currently this can only be done with sufficient know-how about writing a Dockerfile to get additional tools included.

The proposal to simplify adding tools to an existing dev container configuration is to have a `"features"` property listing any additional tools (from a given list of available tools) that should be installed. E.g., the following would add Node.js 14 and the GitHub CLI:
```json
{
	"image": "debian",
	"features": {
		"nodejs": "14",
		"github": true
	}
}
```

A first implementation of this is adding the features by building an additional features image based on the original dev container image. The dev container is then created from this features image. The original image can be from an `"image"`, `"dockerfile"` or `"dockerComposeFile"` configuration.

This issue continues a previously internal discussion.

**Comment by chrmarti** [comment]
Feedback from @Chuxel on the current implementation:
- [ ] We really need text for a hint for the choices that are selected. If you pick Node.js and Java right now, you get two sets of version choices without an indication which one is for which.
- [ ] Since there are options before you get to the selection of features, I think the hint text for the list of checkboxes should be something like "Select additional features to install"
- [ ] Choice/option support is not there for Docker Compose yet which makes the workflow for selecting a version of Python, Node, etc manual still.  I assume that's just not implemented yet.
- [ ] I already mentioned needing the equivalent of `"runArgs": [ "--cap-add=SYS_PTRACE", "--security-opt", "seccomp=unconfined" ]`, but I thought I'd repeat it.
- [ ] I still do find myself wanting to be able to mark certain choices with the option of allowing someone to select "Other..." and let them enter a value. Right now we need to be vague with version numbers and constantly update. I'd love to let people type in a specific version of node for example.
- [ ] Support `${devcontainerName}` substitution in the features.json, e.g., scope volume names to a specific dev container.

**Comment by chrmarti** [comment]
Update:
- Added an optional `"hint"` property for choices that will be shown as the picker's placeholder. E.g., "Select a Node.js version".
- Added `"capAdd"` and `"securityOpt"` properties that take a list of capabilities / security options (implemented for single containers and Docker Compose). E.g.:
```json
"capAdd": [
	"SYS_ADMIN",
	"NET_ADMIN"
],
"securityOpt": [
	"label=user:USER",
	"label=role:ROLE"
]
```

**Comment by Chuxel** [comment]
//cc: @joshspicer @2percentsilk

**Comment by Chuxel** [comment]
@chrmarti One thing I realized was missing from [this list](https://github.com/microsoft/vscode-remote-release/issues/5442#issuecomment-896675211) was the idea of grouping feature options. Two example of this is the `kubectl-helm-debian.sh` script which has three different versions you can specify.  We then have the opposite situation with Java where there's a Gradle and Maven script, but version selection for these should not be presented unless you select Java.  //cc: @joshspicer @2percentsilk

**Comment by chrmarti** [comment]
To keep things simple, I'd prefer to have one version per feature. That also matches what NPM's package.json and the Linux package managers do, they have one version per package. So the `kubectl-helm-debian.sh` script would install 3 features (the script could be split up by feature if that helps).

If one feature depends on another feature, it might makes sense to simply install a suitable version of the dependency if none is specified. So if Gradle is to be installed, but Java is missing, we could just install the latest stable Java version. This again corresponds to what NPM and Linux package managers do and will result in a working installation. The user can later still refine the choice of the Java version.

To avoid overwhelming the user with a very long list of features, we could also move less relevant features to a 'Show All' list, similar to what we do for definitions. E.g., Gradle and Maven could be shown on the initial list for the Java base image, but move to the 'Show All' list for all other base images.

**Comment by Chuxel** [comment]
@chrmarti I don't think that works - features have multiple options. Another example is terraform. There's a TFlint version and TFGrunt version neither of which apply if you're not using Terraform.  Scripts also have multiple options, like Docker-in-Docker supports using Moby or Docker, but again presenting that without first opting into installing Docker doesn't make sense.

This seems to be driving complexity into scripting in a way that is not beneficial to customers.

**Comment by chrmarti** [comment]
I'm not familiar with these particular tools, but if their relation to each other is similar to those I know from the JavaScript world: TFlint and TFGrunt are optional tools depending on Terraform and there might be alternatives like TFhint and TFGulp in the future.

With that in mind, it makes more sense to not nest these optional features, but make them top-level and maybe have their dependencies represented in metadata. I think we need to be careful to not let the current granularity of the scripts drive the schema in the devcontainer.json.

E.g.:

```json
{
  "features": {
    "terraform": "1.0.0",
    "tflint": "0.31.0",
    "tfgrunt": "0.31.8"
  }
}
```

We were also discussing using an array:

```json
{
  "features": [
    {
      "id": "terraform",
      "version": "1.0.0"
    },
    {
      "id": "tflint",
      "version": "0.31.0"
    },
    {
      "id": "tfgrunt",
      "version": "0.31.8"
    }
  ]
}
```

(I like the object notation better since that makes clear that the order doesn't matter and each feature can be added only once. It is also shorter.)

**Comment by stuartleeks** [comment]
If the features are split out, might there not be a case where order _does_ matter? In which case, does the array syntax make sense?

**Comment by stuartleeks** [comment]
Also, the array syntax feels like it could open more options. E.g. for the `az` and `gh` CLIs, I'd like to be able to have an argument for whether or not to mount the host config folder (e.g. the `.azure` folder for `az` - doing this picks up my auth etc from the host):

```json
{
  "features": [
    {
      "id": "terraform",
      "version": "1.0.0"
    },
    {
      "id": "azure-cli",
      "version": "1.0.0",
      "mountHostConfig": true
    },
   ]
}
```

Of course, the object syntax could also be expanded:

```json
{
  "features": {
    "terraform": {
      "version": "1.0.0"
    },
    "azure-cli": {
      "version": "1.0.0",
      "mountHostConfig": true
    },
   }
}
```

**Comment by joshspicer** [comment]
@stuartleeks - i've been tracking what schema changes i've been proposing here, and it's pretty similar to what you're saying here. https://github.com/microsoft/vssaas-planning/issues/5095

@chrmarti  and I talked about it this morning and didn't really land on anything.  I prefer the array syntax for the main reason that it feels more flexible for the future- and given we're not really too firm on a lot of registry details, having a flexible `features` schema would be nice.

**Comment by Chuxel** [comment]
> With that in mind, it makes more sense to not nest these optional features, but make them top-level and maybe have their dependencies represented in metadata. I think we need to be careful to not let the current granularity of the scripts drive the schema in the devcontainer.json.

@chrmarti I think that as we move forward towards Josh's idea of packages, these will be inside the same "feature".

Furthermore, even if they're not, from a UI perspective, I don't want to present options that are not relevant unless they are needed.  We should err on the side of user simplification over tech complexity given the goals here.  So, I'd want this ability in `features.json` even if it doesn't end up in the devcontainer.json that way.

**Comment by chrmarti** [comment]
(re: @stuartleeks)

> If the features are split out, might there not be a case where order _does_ matter? In which case, does the array syntax make sense?

Where the order matters, it should be handled by our implementation. (Similar to how Debian's `apt` takes care of the order.) Leaving the ordering to the user would ask the user for a decision they don't have the context for.

> Of course, the object syntax could also be expanded:

Agreed, also `"azure-cli": "1.0.0"` can be used as the shorthand for using the defaults for any additional options.

(re: @joshspicer)

> @chrmarti and I talked about it this morning. I prefer the array syntax for the main reason that it feels more flexible for the future- and given we're not really too firm on a lot of registry details, having a flexible `features` schema would be nice. My main hesitation, as Stuart pointed out, is if we would ever want to enforce some sort of ordering. Today I can't confidently say one way or another

Could you give any examples of how the registry might require the array syntax? (Using an array for future use without assigning it any semantics now doesn't seem to make sense since by the time we do assign semantics, users would already have their arrays with some more or less random order.)

(re: @Chuxel)

> @chrmarti I think that as we move forward towards Josh's idea of packages, these will be inside the same "feature".

Thinking of features as packages makes me think that we should use a similar granularity as, e.g., Linux packages where each package has a single upstream version.

> Furthermore, even if they're not, from a UI perspective, I don't want to present options that are not relevant unless they are needed. We should err on the side of user simplification over tech complexity given the goals here - which is inherently to abstract while still leaving full power available for those that want it. So, I'd want this ability in `features.json` even if it doesn't end up in the devcontainer.json that way. The feature here is "terraform support" or if we take the Java example, "Java support".

I have the UI in mind and think that we are not limiting ourselves. But we should discuss this here too. I think with the somewhat limiting QuickPick UI we use in VS Code, we might first show a list of features that do not require any other features (e.g., Terraform) and if the user's selection includes any dependencies for additional features, we would then show these additional features in a second list (e.g., TFlint and TFgrunt). This would also manage the length of the first list somewhat.

> As long as we can represent this somewhere in such a way that UI implementations know how to present options, the specifics of how it ends up in devcontainer.json based on selections is less important to the end user. So really my comment here is about the features.json metadata (or whatever that evolves into).

We can tune the metadata towards contributors while we tune the devcontainer.json towards users. E.g., multiple features in the metadata could be handled by the same script (if that helps).

**Comment by Chuxel** [comment]
> I have the UI in mind and think that we are not limiting ourselves. But we should discuss this here too. I think with the somewhat limiting QuickPick UI we use in VS Code, we might first show a list of features that do not require any other features (e.g., Terraform) and if the user's selection includes any dependencies for additional features, we would then show these additional features in a second list (e.g., TFlint and TFgrunt). This would also manage the length of the first list somewhat.

Ultimately, if we have a good UI, what ends up in devcontainer.json is less critical.  Fundamentally, we're invoking installer functions here. Installers provide arguments and there will be more than one.  Let's stick with a simple one.  

With the recent updates to Docker, I can see us needing to provide two options to the installer: Whether to use "Moby" rather than Docker and what version to use. Two options, same feature.  There's similar things in all the scripts about whether to install supplemental tools. Once again, my concern here is passing through the complexity here in a way that does not provide us the ability to group these options in a logical way. We need to provide enough functionality for feature authors (in this case, me) to craft the experience they want to present to the user.

**Comment by chrmarti** [comment]
> > Thinking of features as packages makes me think that we should use a similar granularity as, e.g., Linux packages where each package has a single upstream version.
> > I have the UI in mind and think that we are not limiting ourselves. But we should discuss this here too. I think with the somewhat limiting QuickPick UI we use in VS Code, we might first show a list of features that do not require any other features (e.g., Terraform) and if the user's selection includes any dependencies for additional features, we would then show these additional features in a second list (e.g., TFlint and TFgrunt). This would also manage the length of the first list somewhat.
> 
> @chrmarti These packages are closer to Chocolatey which provide install options than it is to NuGet which backs Chocolatey. The version of the package here is _actually the version of the installer not the thing being installed_. It will be super critical that we do not expect people to create new versions of these packages every time there's an update to the underlying tool/sdk/runtime being installed. But at the same time, these scripts will have breaking changes - the analogy to a GitHub Action here isn't a bad one in this regard.

It's important to keep in mind that there are two audiences: The larger audience are the developers writing a devcontainer.json for their project. We want to keep complexity away from them as much as possible. The smaller audience are the contributors writing their own features with metadata that we currently capture in the internal features.json. We will want to give them as much flexibility as they need to implement their features and they will work with the complexity that comes with that.

> Ultimately, if we have a good UI, what ends up in devcontainer.json is less critical. Fundamentally, we're invoking installer functions here. Installers provide arguments and there will be more than one. Let's stick with some examples.

I think it's important that we keep the devcontainer.json simple as that will be the file most users will work with at some point. There will be many devcontainer.json files in user repositories and we will support existing ones moving forward, so we need to make sure we pick the right amount of flexiblity. The UI we can improve and rewrite any time and we already know that the existing UI using VS Code's QuickPick comes with its advantages thanks to its simplicity, but also its limitations due to that.

> With the recent updates to Docker, I can see us needing to provide two options to the installer: Whether to use "Moby" rather than Docker and what version to use. Two options, same feature. There's similar things in all the scripts about whether to install supplemental tools. Once again, my concern here is passing through the complexity here in a way that does not provide us the ability to group these options in a logical way. Stuart's `mountHostConfig` property for the Azure CLI is another example. I think we need to provide enough functionality for feature authors (in this case, me) to craft the experience they want to present to the user.
> 
> In each of these cases, we'd want to be able to allow someone to select the feature, and then pick options. If that ends up as flat properties in devcontainer.json, ok - though we need to discuss how features actually identify what applies to them.

My proposal is to have:
```json
"azure-cli": "1.0.0"
```

As a shorthand for:
```json
"azure-cli": {
  "version": "1.0.0"
}
```

Allowing for additional options like:
```json
"azure-cli": {
  "version": "1.0.0",
  "mountHostConfig": true
}
```

The version property would be the Azure CLI's version. We can support specifying the installer's version similarly to GitHub Actions, but that should only be needed by CI engineers requiring full control. Most developers will want to use the latest version of the installer.

**Comment by Chuxel** [comment]
@chrmarti Totally good with that. How do we use `features.json` to indicate that the `version` and `mountHostConfig` prompts should only show if the "Azure CLI" feature is selected? That's my point / question.

**Comment by joshspicer** [comment]
I would argue for the `azurecli` case, you generally wouldn't want to specify the version number. I have no idea what version of azure cli I run, but for the most part that doesn't matter.

I still feel 

```json
[
    "azure-cli"
]
```

is cleaner than

```
{
"azure-cli": "XXX"    //Where the user doesn't care about the specific version, which I suspect for a decent chunk of features is true
}

```

It's a bit more verbose if you _do_ want to supply extra config to the feature, but sometimes you won't have to, in which case a single string would do.

Other devcontainer fields like `appPorts`, `mounts`, `runArgs`, etc are arrays.  While perhaps underneath the order _does_ play some effect, we don't advertise order mattering, and people don't get confused.


Re: flexibility

I still also think theres a _possibility_ this could be a valid collection of features in the future:

```json
[
    {
      "id": "node",
      "version": "14"
    },
    {
      "id": "node",
      "version": "16"
    },
    "azure-cli"
]
```

In this case, the object notation restricts us.

**Comment by joshspicer** [comment]
> @Chuxel  @chrmarti Totally good with that. How do we use `features.json` to indicate that the `version` and `mountHostConfig` prompts should only show if the "Azure CLI" feature is selected? That's my point / question.

I think I address that [here](https://github.com/microsoft/vscode-remote-containers/pull/146/files#diff-4c63ec62c537f6293daa9a16e2488bd112d0c3dcf5f1fa6fc21f7fc39374598cR27-R34)

**Comment by Chuxel** [comment]
@joshspicer Azure CLi is just an example of the general idea.  No matter what, each script should include a "latest" version so you don't need to know (and most do already).

**Comment by chrmarti** [comment]
(re: @joshspicer )

> I still also think theres a _possibility_ this could be a valid collection of features in the future:
> 
> ```json
> [
>     {
>       "id": "node",
>       "version": "14"
>     },
>     {
>       "id": "node",
>       "version": "16"
>     },
>     "azure-cli"
> ]
> ```

This opens questions like: Does the order matter? Which version will be on the PATH? How do you switch between versions? Which features support installing multiple versions?

I prefer solving this with existing solutions like `nvm` and if multiple versions of Node.js are required, I would make `nvm` (or any other Node.js version switcher) the feature that optionally takes a list of Node.js versions that should be installed. That would again align features with how Linux packages work, if we go beyond that we need to think about the additional development effort for feature authors.

(re: @Chuxel )

> @chrmarti Totally good with that. How do we use `features.json` to indicate that the `version` and `mountHostConfig` prompts should only show if the "Azure CLI" feature is selected? That's my point / question.

We can add options to each feature, but I wouldn't add feature-like options. E.g., TFlint and TFGrunt should be features, not options to the Terraform feature.

I'm not fully convinced we need options in a first implementation and I don't see the current UI scale to options (I can't imagine users hopping through a growing list of QuickPick steps). So if we support options in metadata and the devcontainer.json, we also want to have defaults for those options, so the UI doesn't have to present them.

**Comment by joshspicer** [comment]
> No matter what, each script should include a "latest" version so you don't need to know (and most do already). @Chuxel 

so each feature, if we want the `latest`, will be added to the devcontainer.json like this (with the syntax @chrmarti  is proposing)?

```json
{
   "azure-cli": "latest",
   "dind": "latest"
}

```

**Comment by chrmarti** [comment]
The object notation is slightly more verbose when you want the latest version. We could also make the array notation shorter when the version is the only additional property by using a `feature@version` (npm-like) or `feature=version` (apt-like) syntax:
```json
[
  "azure-cli",
  "dind",
  "nodejs=14"
]
```

As object:
```json
{
  "azure-cli": "latest",
  "dind": "latest",
  "nodejs": "14"
}
```

Not a big difference. I like the object syntax for indicating that each feature can be installed only once and that the order does not matter. We are using an array for `"extensions"` (and others you point out), there is an argument for consistency here. 🤔

**Comment by joshspicer** [comment]
Are these versions for the _feature "block"_ or the actual _language/tool_ we're installing?

If I say 
```
{
  "nodejs": "latest"
}
```

Does that version signify the version of the:

1. feature (the latest code-change to the script in vscode-dev-containers)

or 
2. The environment variable passed _into_ the script that some  `install_node.sh` script uses to install some version (and in that case is it installing the latest LTS, version 16, something else?)

**Comment by joshspicer** [comment]
I really like the npm/github actions `<package_name>[@version]` syntax, where `@version` is optional.

**Comment by Chuxel** [comment]
> We can add options to each feature, but I wouldn't add feature-like options. E.g., TFlint and TFGrunt should be features, not options to the Terraform feature.

@chrmarti That's a subjective statement.  To me the feature is "Terraform support" - its why we install the extension too. These two CLIs are part of that.  It's not just installing the terraform CLI.

> I'm not fully convinced we need options in a first implementation and I don't see the current UI scale to options (I can't imagine users hopping through a growing list of QuickPick steps). So if we support options in metadata and the devcontainer.json, we also want to have defaults for those options, so the UI doesn't have to present them.

The alternative is to not do anything and wait for complaints. There's no way tflint is a top level feature anymore than pylint or eslint should be. 

We'd have to always install latest and give no options or way to override them and encourage people to "eject" if they need to lock things down more. Even that would be ugly though in its current form.

**Comment by chrmarti** [comment]
@joshspicer 

> Are these versions for the _feature "block"_ or the actual _language/tool_ we're installing?

The actual tool. Most users will be interested in the tool's version and not the feature's version. We can still offer a way to pick the feature's version, but the shorthand should be used for the tool.

Currently each feature declares the valid versions, including 'latest' and 'lts'.

@Chuxel 

> > We can add options to each feature, but I wouldn't add feature-like options. E.g., TFlint and TFGrunt should be features, not options to the Terraform feature.
> 
> @chrmarti That's a pretty subjective statement. To me the feature is "Terraform support" - its why we install the extension too. These two CLIs are part of that. It's not just installing the terraform CLI. This is a bit like saying installing "eslint" and the Eslint extensions is not part of node support. At a certain point, these aren't features anymore since what we're providing is incomplete.

Eslint comes with its own extension and we wouldn't want to force users to have eslint and the eslint extension when all they picked is Node.js.

> > I'm not fully convinced we need options in a first implementation and I don't see the current UI scale to options (I can't imagine users hopping through a growing list of QuickPick steps). So if we support options in metadata and the devcontainer.json, we also want to have defaults for those options, so the UI doesn't have to present them.
> 
> The alternative is to do nothing and wait for complaints. There's no way tflint is a top level feature anymore than pylint or eslint should be. Otherwise then hit a different scale problem with the initial list which is worse since that always appears.

Eslint can be a top-level feature and declare its dependency on the Node.js feature which will allow the QuickPick-based UI to hide it in the initial list and only show it in a second list if the Node.js feature was selected. The UI would work similar if we made it an option to the Node.js feature and other UIs might choose other ways of presenting it.

> We'd have to always install latest and give no options or way to override them and encourage people to "eject" if they need to lock things down more. Even that would be ugly though in its current form. We'd have to document environment variables to set which is pretty kludgey.

We can still offer tool versions without offering additional options for now. We might also offer to specify the feature's version, but I don't see that as a priority and the UI doesn't need to support that since it is an advanced case that can be supported with JSON validation and IntelliSense.

**Comment by Chuxel** [comment]
> Eslint comes with its own extension and we wouldn't want to force users to have eslint and the eslint extension when all they picked is Node.js.

@chrmarti Sounds like we need to book a meeting. We seem to be talking about different concepts at this point.

**Comment by Chuxel** [comment]
Put another way, we should probably agree on the definition of what a "feature" is in this context and how that relates to a package in Josh's thinking. It almost seems like this model says the option is the feature is the package, but I struggle with how we represent options that affect the feature.  Docker-in-Docker's "User Moby" for example.

**Comment by chrmarti** [comment]
I agree, let me try to summarize the two approaches:

1) One feature corresponds to roughly one user-installable tool. E.g., npm would come with the Node.js feature, but yarn would need to be installed separately.
- Fine-grained control also means more configuration.
- The metadata in the features.json declares dependencies. E.g., the yarn feature declares that it depends on the Node.js feature.
- The version property (or a shorthand of it) refers to the tool. We can add a way to configure the feature version.
- The devcontainer.json might look like (or the array equivalent):
```json
{
	"image": "mcr.microsoft.com/vscode/devcontainers/python:3",
	"features": {
		"nodejs": {
			"version": "14",
			"featureVersion": "1.0.1"
		},
		"yarn": "6.0"
	}
}
```
2) One feature corresponds to an entire set of most-often used tools for a topic. E.g., the Node.js feature would also install yarn, eslint, Yeoman and maybe others.
- If any of the tools need priviledged access, the feature needs to declare that also when the user does not need that particular tool.
- Extensions for the tools would be installed also if the user does not use all tools.
- There are potentially multiple version properties, one for each tool and an optional one for the feature.
- The devcontainer.json might look like (or the array equivalent):
```json
{
	"image": "mcr.microsoft.com/vscode/devcontainers/python:3",
	"features": {
		"nodejs": {
			"version": "14",
			"featureVersion": "1.0.1",
			"yarn": "6.0",
			"eslint": "latest",
			"yo": "latest"
		}
	}
}
```

Our initial goal was to make it easier to add simple tools, like a CLI, to an already existing dev container. Both approaches achieve that and are in fact very similar in the simple case (i.e., when there are no dependent tools or options).

Replacing definitions was not a goal. Has this changed?

**Comment by Chuxel** [comment]
> Replacing definitions was not a goal. Has this changed?

@chrmarti Agreed. This is a non-goal. @joshspicer tends to refer to these as a "base", but the term "Dev Container Definition" has been in the wild for long enough I don't think we'd change this externally.  We may be able to reduce the number off them, however.  E.g. the `azure-cli` definition is really `debian` or `ubuntu` plus the checkbox for the Azure CLI.

**Question 1:**

I think what you are describing could work for versions. But there are also options in addition to a version. One in particular that comes to mind is `git-from-src-debian.sh` which we'd likely call just "Git" over time - since building from source is not always required. It has two options of note:

1. Git Version (with latest being one possible  value)
2. Whether a pre-built binary should be used if available

Right now the second option uses a PPA which only works on ubuntu and falls back to building from source otherwise.  In some cases you're not going to trust pre-built binaries and would prefer just to build from source.  Would this be the following?

```json
{
	"image": "mcr.microsoft.com/vscode/devcontainers/python:3",
	"features": {
		"git": {
			"version": "2.33.0",
                        "usePrebuilt": false,
			"featureVersion": "1.0.1"
		}
	}
}
```

**Question 2:**

From a user experience perspective, I don't want developers to have to select Node.js, yarn, and eslint separately from a long list for our definitions - let alone as we scale.

If behind the scenes, Node.js can depend on eslint and yarn, and our UI presents them as "pick Node", then opptions.  How this represents behind the scenes I'm less concerned about.

However, this implies we have "hidden" features.

Is that how this would work?

**Point-of-view:**

I think that question 1 says that features need to have multiple options.  As feature authors, we can then decide whether to group all of them together in one feature, or use separate ones - but that's based on the experience you're trying to deliver for the feature.

**Comment by chrmarti** [comment]
> That said, we should be able to reduce the number off them. E.g. the `azure-cli` definition is really `debian` or `ubuntu` plus the checkbox for the Azure CLI. `docker-in-docker`, `docker-from-docker`, `docker-from-docker-compose` are the same. We'd still want a solid assortment of starting points, but that list can be much smaller - likely centered around images we're pre-building.

Makes perfect sense.

> However, in some cases you're not going to trust pre-built binaries and would prefer just to build from source. This has security benefits and is a popular reaction to supply chain issues that have come to light in recent history. Would this be the following?
> 
> ```json
> {
> 	"image": "mcr.microsoft.com/vscode/devcontainers/python:3",
> 	"features": {
> 		"git": {
> 			"version": "2.33.0",
> 			"usePrebuilt": false,
> 			"featureVersion": "1.0.1"
> 		}
> 	}
> }
> ```

I agree, that's how this would look like. `"usePrebuilt"` and `"featureVersion"` would be optional properties and if they are not needed, a shorthand could be used: `"git": "2.33.0"`.

I still wonder if we need these options in a first implementation. E.g., you could have a `Git` feature that installs the binary when available and falls back to source otherwise and a `Git from Source` features that always installs from source.

> **Question 2:**
> 
> From a user experience perspective, I don't want developers to have to select Node.js, yarn, and eslint separately from a long list for our definitions - let alone as we scale. These are commonly used together enough its worth aggregating them into a single spot in the UX (ignoring the json structure for the moment).
> 
> If behind the scenes, Node.js can depend on eslint and yarn, and our UI presents them as "pick Node", then options, then how this represents behind the scenes I'm less concerned about.

I would say "eslint depends on Node.js", not the other way around. Also: "Node.js is a dependency of eslint". In the metadata it would be eslint listing Node.js as its dependency.

> However, this implies we have "hidden" features.

This depends on the UI. E.g., in the QuickPick UI it is easy to get lost and I would therefore keep the number of options low. So in the QuickPick UI, we could first show a list of features without dependencies as a checkbox list. After the user has picked from those and clicks OK, we check if any of the features with dependencies now have all their dependencies satisfied and present these in a second checkbox list to pick from. I wouldn't necessarily prompt for versions or other options and just configure `"latest"` since that will be the usual choice (e.g. NPM and yarn do something similar when they just write the latest version to the package.json without asking).

This might sound limiting, but I think it is very much targeting our goal to "make it easier to add simple tools, like a CLI, to an already existing dev container".

In the future, a more sophisticated UI could make use of the version and option properties and, e.g., show several features (and maybe dependent features) and their options in a form-like UI that allows for showing more details than the QuickPick UI does.

There is also IntelliSense that we will want to add detailed support for.

> Is that how this would work?

One note: A feature could list more than one other feature as its dependencies.

> **Point-of-view:**
> 
> I think that question 1 says that features need to have multiple options. As feature authors, we can then decide whether to group all of them together in one feature, or use separate ones - but that's based on the experience you're trying to deliver for the feature rather than an inherent limitation of the model.

I agree, this might not always be one or the other. E.g., maybe Moby can be installed alongside Docker and these two can be implemented as two separate features.

> **Addendum:**
> 
> Another point here is that right now we're focusing on single container scenarios, but I think "feature" should be able to spin up a side-car container too. We can then have a "Mongo DB" feature which would eliminate things like `javascript-node-mongodb`. However, I suspect these kinds of features will drive the need for more options beyond just a version - really a sub-point of question 1. This would use the `mongo` image which is going to be significantly faster and less complicated to get up and running and maintain. It's also the more container-y way to do this kind of thing. A good example of an option here is whether mongo should be available at `localhost` or a domain like `db` (affecting network_mode if we used compose here). I don't want us to basically create VMs in containers when we don't have to.

I have so far thought of multi-container scenarios as something that should be covered by Docker Compose because that comes with all the config options one might need and we don't want to replicate these.

**Comment by Chuxel** [comment]
> I still wonder if we need these options in a first implementation. E.g., you could have a Git feature that installs the binary when available and falls back to source otherwise and a Git from Source features that always installs from source.

@chmarti Yes that is how it works now. The other one I don't know how to deal with right now is the "Use Moby" open if we add the ability to specify a Docker version. We'd have to always use Moby as things are right now.

> This depends on the UI.

As long as we can have multiple options, we don't need hidden features.  However, the moment we say there's only one option per feature, and the way to deliver it is through multiple features with dependencies, I'd want to hide some of them that are just there to support a primary feature.

This is a bit like how we have this odd situation right now with "helper extensions" that are visible in the VS Code marketplace. We've discussed wanting to hide them because they do not add sufficient value on their own.  Same thing here.

But, if we can include all these things in one primary feature with options, it's not a problem. 

> I wouldn't necessarily prompt for versions or other options and just configure "latest" since that will be the usual choice ... There is also IntelliSense that we will want to add detailed support for.

One of our problems right now talking to customers is the discoverability of features and even devcontainer.json properties. Suitably motivated people go dig them out, but what we're talking about here is an experience geared towards getting someone up and running quickly.

Either way though, those options have to be defined somewhere so they can get passed into the script to use.

I think it's entirely reasonable to say many of these don't need to show up in the UI, but I they can't be defined at all right now. 
 Even if we had some way to define options that are not exposed in the UI, we'd be in better shape. Right now there's no model to do that since only known variables are passed on as args into the secondary Dockerfile for the build.
 
 > I have so far thought of multi-container scenarios as something that should be covered by Docker Compose because that comes with all the config options one might need and we don't want to replicate these.
 
Agree 100%. However, this is a "take this base and add a DB" variation. Reddis, etc also are examples. Take one of our bases, add another service. Here again, it needs to be our base, and it my be a Docker Compose base (TBD), but you can see what this gets at.

**Comment by chrmarti** [comment]
> > I still wonder if we need these options in a first implementation. E.g., you could have a Git feature that installs the binary when available and falls back to source otherwise and a Git from Source features that always installs from source.
> 
> @chmarti Yes that is how it works now. The other one I don't know how to deal with right now is the "Use Moby" open if we add the ability to specify a Docker version. We'd have to always use Moby as things are right now.

I'm not sure about the difference between using Docker and using "Docker with Moby". Can Moby be installed alongside Docker? Could it be two independent features?

We can add options (beyond just the version) and we don't want artificial features as a workaround. Does the demand for Moby warrant implementing this in the first round?

**Comment by Chuxel** [comment]
> I'm not sure about the difference between using Docker and using "Docker with Moby". Can Moby be installed alongside Docker? Could it be two independent features?

No, you're using one or the other. Either the Docker bits which include additions (e.g. publish to ACI/AWS) but are under a Docker license, or the OSS Moby bits. Both implement the `docker` and `dockerd` commands. Functionally they're very very close, but we use Moby in our images. The script gives users the option to pick today based on their specific needs.

I raised https://github.com/microsoft/vscode-dev-containers/issues/1044 to track what we need to do and the three spots (Kubernetes, Terraform, DinD) that have a challenge and what we could do near term. Day one, if it's a big implementation we can get by, but you can probably see the challenge.

**Comment by joshspicer** [comment]
I am excited to meet and discuss where we want to "officially" draw the line on what _is_ and _isn't_ a feature, nonetheless I think whatever we add in will be flexible enough to support whatever workflow a third-party user would want.  Ie: We (as Microsoft) are going to land on a common set of rules for "what is a feature", but once allow remote, third-party features (and after that enact some sort of registry) this discussion is more about how do we support an individual's scenario, and giving that individual the freedom to create the feature they desire.

----

In reference to the example above illustrating a `version` and `featureVersion`.

```json
// Example from above
{
	"image": "mcr.microsoft.com/vscode/devcontainers/python:3",
	"features": {
		"git": {
			"version": "2.33.0",
			"usePrebuilt": false,
			"featureVersion": "1.0.1"
		}
	}
}
```

My solution to that is to introduce similar syntax to actions with the optional `@<version>`.  

```jsonc
	"features": [
		"python",
		"joshspicer/node@0.0.1": {
                         "version": 14
                  },
		"joshspicer/ghcli"
      ]
```

I am illustrating this with remote features (eg: https://github.com/joshspicer/devcontainer-cli) that can be remotely pulled.  The version indicates the tag of the release.

FYI: I have an implementation of this as a PR on the remote containers repo.

**Comment by chrmarti** [comment]
@joshspicer Note that you've mixed array and object notation in the above example.
Array:
```json
{
	"image": "mcr.microsoft.com/vscode/devcontainers/dotnet:0-5.0",
	"features": [
		"python",
		{
			"id": "joshspicer/node",
			"featureVersion": "0.0.1",
			"version": "14"
		},
		{
			"id": "joshspicer/azcli",
			"version": "3"
		},
		"joshspicer/ghcli@0.0.2"
	]
}
```
Object:
```json
{
	"image": "mcr.microsoft.com/vscode/devcontainers/dotnet:0-5.0",
	"features": {
		"python": "latest",
		"joshspicer/node@0.0.1": "14",
		"joshspicer/azcli": "3",
		"joshspicer/ghcli@0.0.2": "latest"
	}
}
```

When there are additional options, the object syntax would also use an object instead of the version strings as the value.

**Comment by joshspicer** [comment]
> Note that you've mixed array and object notation in the above example.

ah yes, I was mirroring my current implementation in PR (which is dependent on the current schema under the experimental flag).  But I agree the object syntax in this case would use an object as the value instead.

**Comment by chrmarti** [comment]
Update: I have previously added "choice/option support for Docker Compose" and `"customValues": true|false` to the features metadata to allow for other values than the ones supplied as `"values"`.

I have also added support for options while retaining the short-form of only specifying a version. So you can have:
```json
{
	"image": "ubuntu",
	"features": {
		"git": "latest"
	}
}
```

And that is equivalent to:
```json
{
	"image": "ubuntu",
	"features": {
		"git": {
			"version": "latest"
		}
	}
}
```

And to that you can add options like:
```json
{
	"image": "ubuntu",
	"features": {
		"git": {
			"version": "latest",
			"ppa": false
		}
	}
}
```

Currently, for features of type 'choice' the property name is "version", for features of type 'option' the property name is "install" (values true or false). I wonder if we really need the 'option' type or if we should always have a version.

The environment variables used to pass the options (including the version/install property) via features.env to the install script are derived from the feature and option names, e.g.:
```
_BUILD_ARG_GIT_VERSION=latest
_BUILD_ARG_GIT_PPA=true
```

`"buildArg"` is now optional in the feature metadata. I assume we would move over to the automatic ones to keep things simple.

@joshspicer @Chuxel Could you check if this is sufficient for you to start using feature options in your install scripts?

**Comment by joshspicer** [comment]
I'll give it a go and make the edits to the `features.json` in vscode-dev-containers

**Comment by joshspicer** [comment]
so would it not be the case that we'll almost always want `customValues = true`?  Also, when we want to offer a `choice`, aren't we only able to show one "choice" in the UX with the current implemention? What if we want to show several.


Taking the terraform example:
```jsonc
...
},
{
	"id": "terraform",
	"name": "Terraform, tflint, and TFGrunt",
	"type": "option",
	"default": false,
          ...
          ...
},
...
```
With a build script that looking for three different arguments:

```
_VSC_INSTALL_TERRAFORM="terraform-debian.sh ${_VSC_TERRAFORM_VERSION:-latest} ${_VSC_TFLINT_VERSION:-latest} ${_VSC_TERRAGRUNT_VERSION:-latest}"
```

I could imagine us wanting to offer a choice for `TERRAFORM_VERSION`, `TFLINT_VERSION`, AND `TERRAGRUNT `(maybe with `latest` being a default).

With the current implementation, would someone need to know (externally somehow) that these choices exist and can be manually set in the devcontainer?

**Comment by joshspicer** [comment]
> Currently, for features of type 'choice' the property name is "version", for features of type 'option' the property name is "install" (values true or false). I wonder if we really need the 'option' type or if we should always have a version.

I think it's a bit unnecessary to keep the `option` type around.  

I also don't think we should require the user to include a `version`. I know with the object notation we need to have  _something_ passed as the value, but I would think in many cases a user would just want `latest`, in which case asking them to supply that value is redundant.

I would propose:

1. Removing the entire `"type": "(option|choice)"` field.
2. Removing "customValues" field
2. In the feature's `feature.json` entry explicitly state which arguments can be provided (and either explicitly state value, or keep it open-ended)

## Example 1

#### features.json
```jsonc
{
	"id": "terraform",
	"name": "Terraform + friends",
	"choices": {
              version:   [ "lts", "latest","16" ],
              tflint: true,    // Value of true indicates any custom value can be supplied
              terraform: ["none", "lts"]
        },
      ....
      ....

```

#### user's devcontainer
```jsonc
	"features": {
		"terraform": {
			"version": "0.1.2",
			"tflint": 3.1.2,
                         "terragrunt":  4.23.1
		}
	}
```
or
```jsonc
	"features": [
		 {
                      "id": "terraform",
	              "version": "0.1.2",
	              "tflint": 3.1.2,
                      "terragrunt":  4.23.1
		}
	]
```

This is nice since the install script arguments are not going to change, it's still going to require these build args (even if we do default to something if its not present).  This will let us light up the UX, offer intellisense for the devcontainer, and make it clearer what choices are available.

#### features.env

Invoking the script with these build args would look the same, although we'd get a "top-level" `BUILD_ARG_{feature_id}` by default, programatically generated.

```
_BUILD_ARG_TERRAFORM=true  // Added to features.env regardless of which additional "choices" we add. 
_BUILD_ARG_TERRAFORM_VERSION=0.1.2
_BUILD_ARG_TERRAFORM_TFLINT=3.1.2
_BUILD_ARG_TERRAFORM_TERRAGRUNT=4.23.1
```

which could then be invoked like so:
```
if [ ! -z BUILD_ARG_TERRAFORM];then
  "terraform-debian.sh ${_BUILD_ARG_TERRAFORM_VERSION:-latest} ${_BUILD_ARG_TERRAFORM_TFLINT:-latest} ${_BUILD_ARG_TERRAFORM_TERRAGRUNT:-latest}"
fi
```

## Example 2

If a feature didn't want to offer any choices, it could just omit the choices:

#### features.json
```jsonc
{
	"id": "simple",
	"name": "A no-fuss feature",
	"choices": undefined      // or omit completely!
      ....
      ....

```

#### user's devcontainer
```jsonc
	"features": {
		"simple": "true"  // a dummy value that isn't used anywhere
	}
```
or
```jsonc
	"features": [
                 "simple"
	]
```

#### feature.env

```_
BUILD_ARG_SIMPLE=true
```

**Comment by chrmarti** [comment]
@joshspicer I agree we need additional metadata for the options. I'm currently looking into adding features to the JSON-schema (and/or IntelliSense) where we would want additional metadata like a description and type for each option. I'll post an update on that investigation when I have something working end-to-end.

To achieve the goal of making it easy to add tools to an existing dev container, we'll have to make sure all options come with a default value, so all the user has to do is check a box in the UI. IntelliSense can then guide the user through the options.

**Comment by chrmarti** [comment]
I have added metadata for options, currently there are 3 types:
- string with a fixed set of values (an enum).
- string with a set of proposed values, but the user can enter a custom value.
- boolean.

Represented in the metadata as the `"options"` property, e.g.:
```json
		{
			"id": "test",
			"name": "Test",
			"options": {
				"version": {
					"type": "string",
					"proposals": ["latest", "1.0", "2.0"],
					"default": "latest",
					"description": "Version option. Custom value allowed."
				},
				"option1": {
					"type": "boolean",
					"default": true,
					"description": "Boolean option."
				},
				"option2": {
					"type": "string",
					"enum": ["yes", "no", "maybe"],
					"default": "maybe",
					"description": "Enum option."
				}
			}
		}
```

The current implementation still requires `"type"` and `"values"` properties. I think we can replace these with a `"version"` option like in the example above.

I have updated the variables passed to the install script to include one variable for the feature set to "true" and all options set to either their value from the devcontainer.json or their default value from the metadata. E.g.:
```
_BUILD_ARG_TEST=true
_BUILD_ARG_TEST_VERSION=latest
_BUILD_ARG_TEST_OPTION1=true
_BUILD_ARG_TEST_OPTION2=yes
```

To your earlier comment:

> I would propose:
> 
> 1. Removing the entire `"type": "(option|choice)"` field.
> 2. Removing "customValues" field
> 3. In the feature's `feature.json` entry explicitly state which arguments can be provided (and either explicitly state value, or keep it open-ended)
>
> [...]
>
> This is nice since the install script arguments are not going to change, it's still going to require these build args (even if we do default to something if its not present). This will let us light up the UX, offer intellisense for the devcontainer, and make it clearer what choices are available.

I agree, the proposal for `"options"` above should cover this. Additionally we can remove the top-level `"values"` since these can be covered by a `"version"` option.

> #### features.env
> Invoking the script with these build args would look the same, although we'd get a "top-level" `BUILD_ARG_{feature_id}` by default, programatically generated.

Added.

> fyi: I have most implemented this idea of choices here (although i'm sure outdated now): [microsoft/vscode-remote-containers#146](https://github.com/microsoft/vscode-remote-containers/pull/146)

Ack, very similar.

**Comment by chrmarti** [comment]
On the object vs. array notation:

> #### user's devcontainer
> ```js
> 	"features": {
> 		"simple": "true"  // a dummy value that isn't used anywhere
> 	}
> ```
> 
> or
> 
> ```js
> 	"features": [
>                  "simple"
> 	]
> ```

We could always write a version value and use `"latest"` to avoid the artificial `true`.

A few notes:

Array:
- Plus: Shortform with no option.
- Downside: Object with `"id"` and `"${option id}"` for a single option.

Map:
- Upside: Shortform with a single option.
- Upside: Avoids duplicates.
- Upside: Indicates that order does not matter.
- Downside: Repeated `"latest"` (or `true`) value when otherwise no option is needed.

The map/object notation seemed natural when I first wrote `"features"`, but that is somewhat inspired by the package.json's dependencies.

/cc @aeschli

**Comment by jkeech** [comment]
>Upside: Indicates that order does not matter.

I don't think we are going to be able to avoid ordering issues. I believe the solution we need _must_ support the user to force a specific ordering. The array syntax is the simplest to allow the user to pick the order.

Take the following example where the object syntax would not work, but array syntax would: Suppose we have a `python` feature. The user wants to install two different versions of python, so they do something like this:

```
[
    "python": {
        "version": "2.7"
    },
    "python": {
        "version": "3.8"
    }
]
```

Suppose that the feature installs Python and puts it on the front of the `$PATH`. By putting the 3.8 one last, the user can choose to have that one be the default on the `$PATH`. Or if they want 2.7 on the `$PATH`, they put that one last.

You could argue that the user shouldn't be able to specify the same feature twice, and if the feature needs to do multiple things, it needs to handle that internally via options. Even in that case, you could change the example to have one feature called `python-2` and one feature called `python-3`, but that doesn't change the problem. The ordering of which one should be first on the `$PATH` is not something the features or the devcontainer CLI can decide -- it's totally up to the user based on their scenario.

**Comment by Chuxel** [comment]
@chrmarti @jkeech @joshspicer One other design aspect that we should think about here long term is supporting a way to run each feature in a separate layer. I'm having some internal discussions about how we could use cache-from to speed up builds even in the cloud - e.g Actions has some support for this now when you're using buildkit (which we should enable by default anyway).

Right now we're using one layer which gets us by, but at scale and in the cloud, multiple is going to be desirable because you can reuse them.

That said - I do kind of wonder about having a more prescribed ordering because of that. Fewer combinations to cache.

**Comment by chrmarti** [comment]
On the Python example: Debian/Ubuntu and Alpine handle installing Python 2 and 3 by installing `python2` and `python3` executables and (optionally) `python` which in all cases points to Python 2. It seems to make sense to follow the same approach since Linux users are already used to it. Other runtimes, like Node.js, have version switchers, like `nvm`, that we can install. I think following existing approaches and not depending on the order of features in the devcontainer.json will make it easier for users to get a working dev container.

On improving build times: We could split up the `features.env` file (or go back to using Docker's build args) and make each feature install in a layer. That will make it easier to add another feature to an existing image and we can look into using prebuilt base images. Having several definitions with their own base image and features with options will require more prebuilt images to get a good cache hit rate. I suggest we track this as a separate issue we need to better understand before we allow 3rd-parties to contribute features since this will affect how that is done.

**Comment by Chuxel** [comment]
@chrmarti Re: Python, the script builds from source to enable other versions of Python. This is what `pyenv` does as well. It's not fast, however.

**Comment by chrmarti** [comment]
@joshspicer I have updated the extension's code to run everything from `"options"` with a little back compatibility for `"type"`, `"values"`, `"default"`, `"customValues"` and `"hint"`. You can move these over to a `"version"` option, like:
```json
"version": {
	"type": "string",
	"proposals": ["latest", "1.0", "2.0"],
	"default": "latest",
	"description": "Version option."
}
```
(Use `"enum"` instead of `"proposals"` when no custom values are allowed.)

@Chuxel For prebuilt images, building from source for flexibility makes sense and the performance impact is in our CI. For features I'd imagine the default should be to install from prebuilt packages to avoid the delay for building from source on the user's side. If features are used as add-ons (as opposed to the primary technology stack used), using the latest prebuilt package might just be what users expect.

**Comment by chrmarti** [comment]
Previously added `"init"` to the metadata as discussed. Looked into adding `${devcontainerName}` for volume names, the problem there is that volume names cannot use all characters that are allowed in a dev container name and that name can easily change, so a different volume name would result. Maybe we could support using the value from an option for the volume name.

**Comment by Chuxel** [comment]
@chrmarti Just to check `version` the only property supported?

I tried updating the docker-in-docker feature as follows and am not seeing it appear when I ran a quick check. I don't see the "moby" option presented in the UX.

```jsonc
"id": "docker-in-docker",
"name": "Docker (Moby) support (Docker-in-Docker)",
"options": {
    "moby": {
        "type": "boolean",
        "default": true,
        "description": "Install OSS Moby build instead of Docker CE"
    }
},
"entrypoint": "/usr/local/share/docker-init.sh",
"privileged": true,
"init": true,
"containerEnv": {
    "DOCKER_BUILDKIT": "1"
},
// ...
```

I also tried:

```jsonc
//...
"options": {
    "moby": {
        "type": "string",
        "enum": ["true", "false"],
        "default": "true",
        "description": "Install OSS Moby build instead of Docker CE"
    }
},
// ...
```

Similar things happened when I tried multiple properties. e.g. adding a flag as to whether node-gyp dependencies should be installed for the `node` feature in addition to `version`:

```json
"id": "node",
"name": "Node.js (via nvm) and yarn",
"options": {
    "version": {
        "type": "string",
        "proposals": [ "lts", "latest","16", "14", "12" ],
        "default": "latest",
        "description": "Select or enter a Node.js version to install"
    },
    "nodeGypDependencies": {
        "type": "string",
        "enum": [ "true", "false"],
        "default": "true",
        "description": "Install dependencies for compiling native Node.js packages (node-gyp)?"
    }
}
```

**Comment by Chuxel** [comment]
I made updates assuming only `version` is supported in this commit so we can get a build out for testing tomorrow: https://github.com/microsoft/vscode-dev-containers/commit/b99598d7220b216d7bd29f16c180aa3d9b6f1d96

I made a few polish fixes and hit a bug that I resolved.  Take a look @joshspicer @2percentsilk  and see if there's anything else you want to tweak!

**Comment by chrmarti** [comment]
The QuickPick UI only offers "version", only IntelliSense has all the options. Since our goal is to have the checkboxes to get users set up and running, I think we should also remove the "version" picker from the UI and just always use "latest".

**Comment by Chuxel** [comment]
> The QuickPick UI only offers "version", only IntelliSense has all the options.

@chrmarti Got it, so currently we can add them and they'll be picked up in json? 

> Since our goal is to have the checkboxes to get users set up and running, I think we should also remove the "version" picker from the UI and just always use "latest".

@chrmarti I strongly disagree with this part. Our goal is actually to provide a yeoman-like experience in the UI with the ability to do advanced things in json. The problem here again is that people need the option to pin to a version. The way the UI is now with the version picker being available makes that really easy to do which is great - so I don't think we should take this away.  This is a major help in discoverability. It has the same advantages that the settings UI does despite the fact that everything is instrumented in the json editor too.

Also, as we add in remote features, will those work in the json editor?

I'd like to expose additional options in the UI, but I'd like to hear @2percentsilk's thoughts here too.

**Comment by chrmarti** [comment]
@Chuxel Correct, additional options will be available in JSON and suggested in IntelliSense.

I see you've moved all `"type": "choice"` features over, can we also move the `"type": "option"` ones over to offer a single `"latest"` version? (Or more versions where applicable.) That will allow each feature to have the `"feature_id": "latest"` shorthand and will make it easy to add versions in the future.

**Comment by joshspicer** [comment]
> @Chuxel Correct, additional options will be available in JSON and suggested in IntelliSense.
> 
> I see you've moved all `"type": "choice"` features over, can we also move the `"type": "option"` ones over to offer a single `"latest"` version? (Or more versions where applicable.) That will allow each feature to have the `"feature_id": "latest"` shorthand and will make it easy to add versions in the future.

I'll move these over to make sure I have a good grasp on the new syntax

**Comment by joshspicer** [comment]
> ...can we also move the `"type": "option"` ones over to offer a single `"latest"` version? (Or more versions where applicable.) That will allow each feature to have the `"feature_id": "latest"` shorthand and will make it easy to add versions in the future.

Is it required for one of the options to be `version`?  We have no concept of version for some scripts (like `docker-in-docker`)

**Comment by joshspicer** [comment]
Please let me know if anything else should be tweaked in the `features.json` ^

I added in a `version` option for all of the features that didn't have it, with a description of "Not used" when applicable.

**Comment by Chuxel** [comment]
@chrmarti lol. As if to accentuate the importance of version selection, we've had two breaks today that could be solved with pinning to a specific version.

1. One was Docker Compose, though we don't expose that version in the script and probably should. 
 Here again, people expect this to be there if Docker is installed. https://github.com/microsoft/vscode-dev-containers/issues/1075
 2. The other is there have been some issues with the latest terragrunt release. https://github.com/gruntwork-io/terragrunt/issues/1829, https://github.com/gruntwork-io/terragrunt/issues/1832
 
 We're finding these in CI, which is great, but if we don't have the ability to recommend people pin to specific releases rather than "latest", this will keep happening.  (We also need to recommend  pre-building images since that side-steps this entirely, but that's a separate topic.)

**Comment by chrmarti** [comment]
> Is it required for one of the options to be `version`? We have no concept of version for some scripts (like `docker-in-docker`)
> 
> I suppose the shorthand then would be broken like you're mentioning. I guess that means authors _need_ to add in a version to their feature's features.json entry? Seems like an unnecessary restriction IMO

@joshspicer `"version"` is optional and `"feature_id": true` should continue to work for these features.

**Comment by chrmarti** [comment]
@joshspicer I see a few cases (e.g., PowerShell) where a version makes sense even if we currently don't offer any but the "latest". For these I recommend keeping the version, so we can add versions in the future without changing the schema from boolean to string.

**Comment by chrmarti** [comment]
Available without experimental setting with Remote-Containers 0.200.0. Closing this issue, let's split out new issues for the remaining tasks.
